### DISCUSSION TITLE
Is there a better way to integrate with custom phone auth provider?

### DISCUSSION THREAD

**Discussion by gerynugrh** [supabase/supabase#14774] [Questions] [answered]
For some users in ID, integrating phone auth with existing provider (Twilio, Message Bird) is not preferable given the high price per sms compared to other provider. Hence, currently we do this by using the following trigger

```sql
DROP TRIGGER IF EXISTS trigger_otp ON auth.users;

CREATE OR REPLACE FUNCTION trigger_otp()
RETURNS TRIGGER
AS $$
DECLARE otp bigint;
BEGIN
    IF OLD.confirmation_token = NEW.confirmation_token THEN
      RETURN NEW;
    END IF;
    IF NEW.confirmation_token = '' THEN
      RETURN NEW;
    END IF;
    otp := FLOOR(RANDOM() * 900000 + 100000);
    INSERT INTO public.tmp_otp_trigger (code, user_id, phone, confirmation_token) values (otp, NEW.id, NEW.phone, NEW.confirmation_token);
    NEW.confirmation_token := encode(sha224(concat(NEW.phone,otp)::bytea), 'hex');
    NEW.confirmation_sent_at := now();
    RETURN NEW;
END;
$$ LANGUAGE plpgsql SECURITY definer;

CREATE TRIGGER trigger_otp BEFORE INSERT OR UPDATE ON auth.users FOR EACH ROW EXECUTE PROCEDURE trigger_otp();
```

It works by generating another random OTP, hash, and then update `auth.users` while also inserting row into another temporary table for OTPs. Later, there will be an edge function that triggers when a record is inserted into `tmp_otp_trigger`.

We use this in our apps and it's working well, but is there a better way to do this that we haven't notice yet? Also side question, do you happen to know how to generate otps other than using pgsql `random()` function (given that this function is not cryptographically secure)?

**Comment by ZelimDamian** [comment]
@gerynugrh We are having a similar issue and have been struggling to make it work with a different SMS gateway. Could you provide more information about how the trigger above fits into the Supabase auth so we could maybe replicate?

**Reply by gerynugrh** [reply]
Hi @ZelimDamian I got the inspiration from this post
https://github.com/orgs/supabase/discussions/5358

In short, when we call `signInWithOtp` in client, supabase will create a new record in `auth.users` table. The table will contain the phone number the user intend to send the OTP to and `confirmationToken`. 

Supabase use `confirmationToken` during verify call by hashing the OTP given by user and the phone number and check whether the hash generated by user request is equal to `confirmationToken` stored in db. The hash formula should be similar to this
```
encode(sha224(concat(NEW.phone,otp)::bytea), 'hex');
```

Due to above logic, it should be feasible to replace the `confirmationToken` and trick supabase so they think that the OTP sent to user is a different OTP, generated by pgsql function above. Hence, we just need to pass the newly generated OTP to a supabase edge function that can use any sort of SMS OTP implementation underneath.

**Reply by ZelimDamian** [reply]
Thank you for the explanation @gerynugrh ! Will try to implement the same as our current approach to treat signed phonenumbers as passwords feels quite off.

**Reply by nadirbelarouci** [reply]
@gerynugrh WHy don't you trigger the edge function from the `auth.users` table right away ? why the tmp table is necessary ?

**Reply by gerynugrh** [reply]
Hi @nadirbelarouci I would need to modify the confirmation token expected from `auth.users`. This requires me to modify `auth.users` after an update is happening. As far as I'm aware, I can't modify `auth` schema in supabase from an edge function, hence I need to do it via db trigger.

**Reply by nadirbelarouci** [reply]
Hi @gerynugrh, I am still missing some context, why would you change the confirmation token ? isn't this done by the trigger: `NEW.confirmation_token := encode(sha224(concat(NEW.phone,otp)::bytea), 'hex');`
Another question 😄 
Will this prevent sending the sms from default providers ?

**Reply by gerynugrh** [reply]
Hi, I haven't worked on this problem for some time. But if my memory servers correctly it was because supabase only store the hased confirmation token. Hence there is no way for us to know the OTP that supabase sent to the SMS provider from the db record.

If we want to send an SMS OTP to another provider, we need to know the OTP code that supabase expect, but this impossible by reading the newly inserted `auth.users` record alone. Hence what we're doing is generate a new OTP, update `auth.users` to force supabase to acknowledge the newly generated OTP and send the OTP to another SMS provider.

**Comment by lucaspeyrin** [comment]
Any updates on this ? Is there a way to have a custom SMS provider ?

**Reply by 4lirexa** [reply]
check out https://medium.com/@lavisht22/optimizing-sms-authentication-costs-with-supabase-a-deep-dive-4a2f1b4a1a68

**Comment by AliKales** [comment]
https://medium.com/@lavisht22/optimizing-sms-authentication-costs-with-supabase-a-deep-dive-4a2f1b4a1a68
this is working very well but do not use Twilio as your provider otherwise it will fail just choose any other one

**Reply by ewan133** [reply]
hey is it possible to use it for flutter i'm new  to app development and i have many questions regarding how to integrate it

**Comment by LeoVS09** [comment, marked as answer]
For anyone who, like me, spends much time digging into guides on how to implement it. Supabase actually already has custom SMS provider support, through the [Send SMS hook](https://supabase.com/docs/guides/auth/auth-hooks/send-sms-hook).

**Comment by rayasoren** [comment]
Disclosure up front: I work on MyOTP, so read this as a vendor answer.

@LeoVS09 is right that the Send SMS hook is the supported route now, and the part worth emphasising for anyone landing here from a search: the hook takes **any HTTPS endpoint**. You are not limited to providers that have a Supabase integration, which is what the original question was really blocked on.

On the cost problem that started this thread — the thing that usually dominates the bill is not the SMS, it is the per-verification platform fee layered on top of it. Twilio Verify lists $0.05 per successful verification *plus* roughly $0.0083 for a US message. If you send the SMS yourself through any carrier API you skip that fee entirely, and that is most of the difference people notice when they switch. Worth pricing your own volume both ways before assuming a provider swap is the fix.

Two things that bit us building against this hook, useful regardless of who you end up using:

- The hook payload contains the OTP code itself, so whatever you point it at is now in scope for however you handle secrets. Don't log the payload verbatim.
- Supabase considers the send successful if your endpoint returns 2xx. That is acceptance, not delivery. If you care whether the code actually reached a handset you need delivery reporting from the provider on top, otherwise a silently failing route looks identical to a working one.

We implemented the hook endpoint directly and handle routing, retries and delivery reporting behind it, with WhatsApp and Telegram on the same call for countries where SMS is expensive or unreliable. Happy to answer setup questions here whichever provider you pick — the hook is the genuinely useful part of this.
